### DISCUSSION TITLE
Best practices for handling server offline (i.e. Network Error)

### DISCUSSION THREAD

**Discussion by rsdickerson** [urql-graphql/urql#1476] [Q&A] [answered]
I have implemented the OfflineExchange and have a query using requestPolicy: "cache-and-network". If I start my app with the server on-line to first populate the cache and then stop the server I'm seeing the query response come from the cache (as it should) and the network request failing without bubbling up the "Network Error". So, in this scenario using the OfflineExchange is working great.

But, what if I still want to know that the query response came from the cache and is not fresh so that I can notify the user that the data they are viewing is stale? What would be the best way to do that?

**Comment by kitten** [comment, marked as answer]
I'd combine some logic to detect offline states, an example of which you can find in the source itself: https://github.com/FormidableLabs/urql/blob/65e15c568c9a0755894dfb02995f0262e1fe75f9/exchanges/graphcache/src/offlineExchange.ts#L59
With the `errorExchange` which can notify you of any error.

This should allow you to detect temporary offline states and notify the user of them. Alternatively, a lot of web apps also just use the window online/offline events.

While these events aren't as accurate (and will in your case not fire) they often can be used to switch an app into the "offline" state without requiring a request first. The disadvantage of only relying on these though is that they may not fire if a network device still has an apparently working connection that just isn't transferring (think of a broken 3G connection that has signal but where the network is overloaded)

**Reply by rsdickerson** [reply]
Thanks. I'll check out the errorExchange.

The app is currently monitoring online/offline which I use to disable some features when offline. That's working fine when the device loses its network connection. I'm mostly just trying to handle when the server is down or your example of the broken 3G connection.

I had forgotten about how the useQuery() hook can return "stale". If I understand the documentation correctly, stale would be set to true when the response is fulfilled from cache and set to false when the response is from the network request. I haven't tried it yet, but I'm assuming since when using the OfflineExchange the Network Error just gets eaten by urql that stale would remain true. Is that correct?

**Reply by kitten** [reply]
It shouldn't be set. The reasoning here being that if you get a result that's partial (or complete) from the cache then it'll be entirely offline and valid.

So it's still necessary to detect whether the app is offline separately.

The idea is that an offline-first application should be able to continue working without exposing temporary offline states, since unstable connections can often lead to very temporary delays that the `retryExchange` can smooth over.

**Reply by rsdickerson** [reply]
For anyone needing to add an errorExchange, it is documented in the [Exchange Author Guide](https://formidable.com/open-source/urql/docs/advanced/authoring-exchanges/#using-exchanges).
